packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts

```diff
@@ -187,27 +195,39 @@ export class FederatedQuadSource implements IQuadSource {
   }
 
   public match(subject: RDF.Term, predicate: RDF.Term, object: RDF.Term, graph: RDF.Term): AsyncIterator<RDF.Quad> {
-    // Counters for our metadata
-    const metadata: MetadataQuads = { cardinality: { type: 'exact', value: 0 }, canContainUndefs: false };
-    let remainingSources = this.sources.length;
-
-    // Anonymous function to handle cardinality from metadata
-    const checkEmitMetadata = (currentTotalItems: number, source: IDataSource,
-      pattern: RDF.BaseQuad | undefined, lastMetadata?: MetadataQuads): void => {
-      if (this.skipEmptyPatterns && !currentTotalItems && pattern && !this.isSourceEmpty(source, pattern)) {
-        this.emptyPatterns.get(source)!.push(pattern);
-      }
-      if (!remainingSources) {
-        if (lastMetadata && this.sources.length === 1) {
-          // If we only had one source, emit the metadata as-is.
-          it.setProperty('metadata', lastMetadata);
-        } else {
-          it.setProperty('metadata', metadata);
+    // A function that is called every time metadata is found,
+    // and will accumulate all metadata if the metadata from all sources have been defined.
+    const tryAccumulatingMetadata = async(): Promise<void> => {
+      // Only accumulate when the metadata of all sources have been defined
+      if (accumulatingMetadata.size === this.sources.length) {
+        // Accumulate metadata using mediator
+        let accumulatedMetadata: MetadataQuads = <MetadataQuads> (await this.mediatorRdfMetadataAccumulate
+          .mediate({ mode: 'initialize', context: this.contextDefault })).metadata;
+        for (const appendingMetadata of accumulatingMetadata.values()) {
```

Comment:
Does `accumulatingMetadata` offer a method like `forEach` instead. I'm assuming `values()` creates an array which means that all items have to be materialized before iterating at the moment